cpp/src/lexer/lexer.cc

```diff
@@ -0,0 +1,67 @@
+#include <cctype>
+
+#include "lexer.hh"
+
+Lexer::Lexer(std::string_view input)
+: input_{input}
+, length_{input.length()}
+{
+}
+
+inline bool isWhitespace(char ch)
+{
+	return ch == ' ' || ch == '\t' || ch == '\r' || ch == '\n';
+}
+
+Token Lexer::nextToken() noexcept
+{
+	// skip whitespace
+	while (position_ < length_ && isWhitespace(input_[position_]))
+		position_++;
+
+	// check if we're already off the end
+	if (position_ >= length_)
+		return { TokenType::eof };
+
+	// peek the next character
+	const auto start_position = position_++;
+	const auto ch = input_[start_position];
+
+	// test for single-character tokens
+	switch (ch) {
+		case '=': return { TokenType::assign    };
+		case ';': return { TokenType::semicolon };
+		case '(': return { TokenType::lparen    };
+		case ')': return { TokenType::rparen    };
+		case ',': return { TokenType::comma     };
+		case '+': return { TokenType::plus      };
+		case '{': return { TokenType::lsquirly  };
+		case '}': return { TokenType::rsquirly  };
+	}
+
+	// parse integer
+	if (isdigit(ch)) {
+		while (position_ < length_ && isdigit(input_[position_]))
```

Comment:
This (and the similar lines below) looks like a good place to use string_view::first_not_of.

Reply:
i very much doubt that's more efficient that `is*()`

Reply:
new version uses a `string_view` and  a `string_view::iterator` only.